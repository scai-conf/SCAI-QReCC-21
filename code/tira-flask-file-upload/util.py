def vm_name(request):
    if not(request) or not(request.headers) or not('X-Disraptor-Groups' in request.headers):
        return None
    
    groups = request.headers.get('X-Disraptor-Groups')
    if not groups or 'tira_vm_' not in groups:
        return None
    ret = (',' + groups +',').split(',tira_vm_')[1].split(',')[0]
    ret = ret if ret else None

    print('Handle request for virtual machine "' + ret + '".')
    return ret

def run_id(current_time):
    return current_time.strftime('%Y-%m-%d-%H-%M-%S')

def next_software_num(vm_id):
    from pathlib import Path
    if Path('/mnt/ceph/tira/model/softwares/scai-qrecc/' + vm_id + '/softwares.prototext').is_file():
        max_num = 0
        with open('/mnt/ceph/tira/model/softwares/scai-qrecc/' + vm_id + '/softwares.prototext', 'r') as f:
            for l in f:
                if 'count: "' in l:
                    l = l.split('count: "')[1]
                    if '"' in l:
                        l = int(l.split('"')[0])
                        if l > max_num:
                            max_num = l
        return max_num +1
    return 1

def run_prototext(run_id, input_dataset, software_id):
    return '''softwareId: "software''' + str(software_id) + '''"
runId: "'''+ run_id + '''"
inputDataset: "''' + input_dataset + '''"
inputRun: "none"
downloadable: true
deleted: false
taskId: "scai-qrecc"
accessToken: "manual-run-no-access-token"'''

def run_software(current_time, dataset, num, run_name):
    current_time = current_time.strftime('%a %b %d %H:%M:%S UTC %Y')
    return '''softwares {
  id: "software''' + str(num) + '''"
  count: "''' + str(num) + '''"
  command: "# This software documents a manual upload of a run file named ''' + run_name + ''' on ''' + current_time + '''"
  workingDirectory: ""
  dataset: "''' + dataset + '''"
  run: "none"
  creationDate: "''' + current_time + '''"
  lastEditDate: "''' + current_time + '''"
  deleted: false
}
'''

def build_run(data, vm, current_time, input_dataset, run_name):
    from subprocess import check_output
    import os
    
    out_dir = '/mnt/ceph/tira/data/runs/' + input_dataset
    run_dir = os.path.join(out_dir, vm, run_id(current_time))
    os.makedirs(os.path.join(run_dir, 'output'), exist_ok=True)
    os.makedirs(os.path.join('/mnt/ceph/tira/model/softwares/scai-qrecc/', vm), exist_ok=True)
    output_file = os.path.join(run_dir, 'output', 'run.json')

    with open(os.path.join(run_dir, 'output', 'run.json'), 'wb') as f:
        f.write(data)

    with open(os.path.join(run_dir, 'run.prototext'), 'w') as f:
        f.write(run_prototext(run_id(current_time), input_dataset, next_software_num(vm)))

    with open(os.path.join(run_dir, 'file-list.txt'), 'wb') as f:
        file_list = check_output(['tree', '-ahv', os.path.join(run_dir, 'output')])
        f.write(file_list)

    with open(os.path.join('/mnt/ceph/tira/model/softwares/scai-qrecc/', vm, 'softwares.prototext'), 'a+') as f:
        f.write('\n\n' + run_software(current_time, input_dataset, next_software_num(vm), run_name))

    with open(os.path.join(run_dir, 'stdout.txt'), 'wb') as f:
        f.write('This software was not executed in TIRA and documents a manual upload of a run file named "' + run_name + '" on ' + current_time.strftime('%a %b %d %H:%M:%S UTC %Y') + '.\n')

    with open(os.path.join(run_dir, 'size.txt'), 'wb') as f:
        f.write(check_output(['bash', '-c', '(du -sb "' + run_dir + '" && du -hs "' +  run_dir + '") | cut -f1']))
        f.write(check_output(['bash', '-c', 'find "' + os.path.join(run_dir, 'output') + '" -type f -exec cat {} + | wc -l']))
        f.write(check_output(['bash', '-c', 'find "' + os.path.join(run_dir, 'output') + '" -type f | wc -l']))
        f.write(check_output(['bash', '-c', 'find "' + os.path.join(run_dir, 'output') + '" -type d | wc -l']))

    check_output(['chmod', '775', '-R', os.path.join(out_dir, vm)])

