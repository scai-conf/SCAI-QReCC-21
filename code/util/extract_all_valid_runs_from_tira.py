#!/usr/bin/env python3

# Manually collected
VALID_RUNS = {
    'scai-qrecc21-test-dataset-2021-07-20': {
        'rachael': ['2021-09-15-09-06-44', '2021-09-15-09-05-06', '2021-09-08-21-49-44', '2021-09-08-07-07-57', '2021-09-15-09-08-40', '2021-09-08-15-40-34', '2021-09-08-07-09-57', '2021-09-15-09-07-49', '2021-09-04-10-38-07'],
        'scai-qrecc21-basic-baseline': ['2021-08-26-17-46-42'],
        'scai-qrecc21-gpt3-baseline': ['2021-07-21-08-56-42'],
        'scai-qrecc21-simple-baseline': ['2021-07-20-21-01-58'],
        'torch': ['2021-09-20-05-42-29'],
        'ultron': ['2021-09-04-17-16-58'],
    },
    'scai-qrecc21-test-dataset-rewritten-2021-07-20': {
        'rachael': ['2021-09-04-10-39-42', '2021-09-15-19-36-31', '2021-09-06-09-21-43'],
        'rali-qa': ['2021-09-09-13-01-07'],
        'scai-qrecc21-basic-baseline': ['2021-08-26-14-42-21'],
        'scai-qrecc21-simple-baseline': ['2021-07-21-08-58-25'],
        'ultron': ['2021-09-08-15-08-00', '2021-09-19-18-40-28', '2021-09-08-15-04-28', '2021-09-08-15-07-30', '2021-09-13-17-56-51', '2021-09-13-17-47-43'],
    },
}


def parse_run_name(file_name):
    import json
    try:
        lines = [i for i in open(file_name)]
        ret = json.loads(lines[-1])['run_name']
        if not ret:
            return 'no-run-identifier'
        return ret
    except:
        return 'no-run-identifier'


def copy(src, target):
    import shutil
    shutil.copy(src, target)

if __name__ == '__main__':
    from tqdm import tqdm
    for test_dataset in VALID_RUNS:
        for user in tqdm(VALID_RUNS[test_dataset]):
            for run_timestamp in VALID_RUNS[test_dataset][user]:
                run_dir = '/mnt/ceph/tira/data/runs/' + test_dataset + '/' + user + '/' + run_timestamp + '/'
                run_name = parse_run_name(run_dir + 'stdout.txt')
                copy(run_dir + 'output/run.json', 'out/' + test_dataset + '/' + test_dataset + '-' + user + '-' + run_name  + '-' + run_timestamp + '.json')

