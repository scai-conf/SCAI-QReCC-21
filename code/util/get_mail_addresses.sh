#!/bin/bash
cat /dev/stdin | grep -o "[^,]*@[^,]*"
