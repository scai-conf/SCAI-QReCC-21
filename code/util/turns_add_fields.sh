#!/bin/bash

# Usage: <turns_file> <annotations_file>
# Where:
#   - <turns_file>
#       JSON list of objects, each of which has a "Conversation_no" and
#       "Turn_no" field with integer values
#   - <annotations_file>
#       Each line: <conversation_no> <turn_no> <addition (one or more)>

turns_file=$1
additions_file=$2

cat $turns_file \
  | jq --indent 4 . \
  | awk 'FILENAME == "'$additions_file'" {
      conversation_no = $1
      turn_no = $2
      addition = $3
      for (f = 4; f <= NF; ++f) {
        addition = addition" "$f
      }
      additions[conversation_no"_"turn_no] = addition
    } FILENAME == "/dev/stdin" { 
      if ($1 == "\"Conversation_no\":") {
        conversation_no = $2
        sub(/,/, "", conversation_no)
      } else if ($1 == "\"Turn_no\":") {
        turn_no = $2
        sub(/,/, "", turn_no)
      } else if ($0 ~ /^    }/) {
        if (conversation_no"_"turn_no in additions) {
          printf ", %s", additions[conversation_no"_"turn_no]
        }
        conversation_no = ""
        turn_no = ""
      }
      print $0
    }' $additions_file /dev/stdin \
  | jq --indent 4 --sort-keys .
