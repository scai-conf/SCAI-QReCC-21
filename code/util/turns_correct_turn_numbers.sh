#!/bin/bash

# Usage: <turns_file>
# Where:
#   - <turns_file>
#       JSON list of objects, each of which has a "Conversation_no" and
#       "Turn_no" field with integer values

turns_file=$1
cat $turns_file \
  | awk '{
      if ($1 == "\"Conversation_no\":") {
      old_conversation_no = conversation_no
      conversation_no = sprintf("%d", $2)
      }

      if ($1 == "\"Turn_no\":") {
        if (conversation_no == old_conversation_no) {
          turn_no = turn_no + 1
        } else {
          turn_no = 1
        }
        $2 = sprintf("%d", turn_no)
      }
      print $0
    }' \
  | jq '.'
