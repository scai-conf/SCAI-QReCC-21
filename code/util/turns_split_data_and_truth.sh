#!/bin/bash

turns_file=$1
output_dir=$2
mkdir -p $output_dir

max_conversation_no=$(cat $turns_file | jq '.[].Conversation_no' | awk '$1>max{max=$1}END{print max}')
echo "Max Conversation_no: $max_conversation_no"
cat $turns_file \
  | jq --compact-output --sort-keys '.[]' \
  | gawk 'BEGIN {
      output_question = "jq --indent 4 --sort-keys --slurp . > '$output_dir'/questions.json"
      output_truth = "jq --indent 4 --sort-keys --slurp . > '$output_dir'/ground-truth.json"
      next_conversation_no = '$max_conversation_no' + 1
    } function add_question(conversation_no, turn_no, context, question) {
      printf "{\"Conversation_no\":%d,\"Turn_no\":%d,\"Context\":%s,\"Question\":%s}\n", conversation_no, turn_no, context, question | output_question
    } function add_truth(model_conversation_no, model_turn_no, original_conversation_no, transformer_conversation_no, human_conversation_no, truth_rewrite, truth_passages, truth_answer) {
      printf "{\"Turns\":{\"model\":{\"Conversation_no\":%d,\"Turn_no\":%d},\"original\":{\"Conversation_no\":%d,\"Turn_no\":1},\"transformer\":{\"Conversation_no\":%d,\"Turn_no\":1},\"human\":{\"Conversation_no\":%d,\"Turn_no\":1}},\"Truth_rewrite\":%s,\"Truth_passages\":%s,\"Truth_answer\":%s}\n", model_conversation_no, model_turn_no, original_conversation_no, transformer_conversation_no, human_conversation_no, truth_rewrite, truth_passages, truth_answer | output_truth
    } {
      match($0, /.*,"Context":(\[.*\]),"Conversation_no":/, contexts); context = contexts[1]
      match($0, /.*,"Conversation_no":([0-9]*),/, conversation_nos); conversation_no = conversation_nos[1]
      match($0, /.*,"Question":(".*"),"Transformer_rewrite"/, questions); question = questions[1]
      if (question == "") {
        match($0, /.*,"Question":(".*"),"Truth_answer"/, questions); question = questions[1]
        transformer_rewrite = question
      } else {
        match($0, /.*,"Transformer_rewrite":(".*"),"Truth_answer"/, transformer_rewrites); transformer_rewrite = transformer_rewrites[1]
      }
      match($0, /.*,"Truth_answer":(".*"),"Truth_passages"/, truth_answers); truth_answer = truth_answers[1]
      match($0, /.*,"Truth_passages":(\[.*\]),"Truth_rewrite":/, truth_passagess); truth_passages = truth_passagess[1]
      match($0, /.*,"Truth_rewrite":(".*"),"Turn_no"/, truth_rewrites); truth_rewrite = truth_rewrites[1]
      match($0, /.*,"Turn_no":([0-9]*)/, turn_nos); turn_no = turn_nos[1]

      add_question(conversation_no, turn_no, context, question)
      if (turn_no == 1) {
        add_truth(conversation_no, turn_no, conversation_no, conversation_no, conversation_no, truth_rewrite, truth_passages, truth_answer)
      } else {
        original_conversation_no = next_conversation_no++
        add_question(original_conversation_no, 1, "[]", question)

        transformer_conversation_no = original_conversation_no
        if (transformer_rewrite != question) {
          transformer_conversation_no = next_conversation_no++
          add_question(transformer_conversation_no, 1, "[]", transformer_rewrite)
        }

        human_conversation_no = original_conversation_no
        if (truth_rewrite != question) {
          if (truth_rewrite != transformer_rewrite) {
            human_conversation_no = next_conversation_no++
            add_question(human_conversation_no, 1, "[]", truth_rewrite)
          } else {
            human_conversation_no = transformer_conversation_no
          }
        }

        add_truth(conversation_no, turn_no, original_conversation_no, transformer_conversation_no, human_conversation_no, truth_rewrite, truth_passages, truth_answer)
      }
    } '

