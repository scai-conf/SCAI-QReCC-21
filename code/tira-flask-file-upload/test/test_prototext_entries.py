from util import run_software, run_prototext, run_id
from approvaltests.approvals import verify
from datetime import datetime

def test_run_prototext():
    current_time = datetime.strptime('Jun 1 2005  1:33PM', '%b %d %Y %I:%M%p')
    verify(run_prototext(run_id(current_time), 'scai-qrecc21-test-dataset-2021-05-15', 1))

def test_run_software_1():
    current_time = datetime.strptime('Jun 1 2005  1:33PM', '%b %d %Y %I:%M%p')
    verify(run_software(current_time, 'my-test-dataset-1', 23, 'my-run-tag'))

