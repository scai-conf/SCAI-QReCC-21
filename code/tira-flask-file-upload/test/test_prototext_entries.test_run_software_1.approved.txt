softwares {
  id: "software23"
  count: "23"
  command: "# This software documents a manual upload of a run file on Wed Jun 01 13:33:00 UTC 2005"
  workingDirectory: ""
  dataset: "my-test-dataset-1"
  run: "none"
  creationDate: "Wed Jun 01 13:33:00 UTC 2005"
  lastEditDate: "Wed Jun 01 13:33:00 UTC 2005"
  deleted: false
}
